ls1intum/Artemis
# `Development`: Improve notification performance

### Checklist
#### General


#### Server
#### Client

### Motivation and Context
We found out that the `notifications` DB call on production is currently very slow and unfortunately gets triggered by all participants at the same time at the end of the retreat, because then the sidebar is visible again. What we have implemented:
1. Workaround so that only notifications for the last 30 days are displayed (this makes the DB Call significantly faster) --> will be reworked in a follow-up.
2. If the sidebar is shown again after it was already visible (e.g. at the end of the exam) the REST call should **NOT** be executed.
3. When reloading the application (e.g. browser refresh) the notifications should be loaded only after a delay of 15s, so that with frequent reloads no overload comes at the server by it
4. New indices that reduce the system notifications DB call from 100ms to 5ms.

**Important: The notifications should in no case be able to overload the database, because they are never that important. :bangbang:**

Changes:
- Workaround that notifications are only shown for the last 30 days --> will be reworked in a follow up
- If the sidebar is reenabled after being shown previously (e.g. after the exam conduction) it should not resend the REST call to fetch the notifications again.
- During page reload, notifications should only be loaded after 15 seconds --> follow up for the case that the student clicks on the sidebar before these 15 seconds
- New indices for the data base

### Steps for Testing

Prerequisites:
- 1 Instructor
- 1 Student

1. Log in to Artemis
2. Check that you can see your notifications after 15 seconds as before
3. Check that you can receive new notifications (e.g. rescheduling of your tutorial group)

#### Exam Mode Testing



Prerequisites:
- 1 Instructor
- 1 Students
- 1 Exam

1. Log in to Artemis
2. Check that the notifications are not reloaded after you submitted your exam

### Review Progress






#### Performance Review
#### Code Review
#### Manual Tests
#### Exam Mode Test

### Test Coverage





#### Client

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|---------------------:|

#### Server

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|---------------------:|
| [TutorialGroupService.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5290/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service.tutorialgroups/TutorialGroupService.html) | 95% | ✅ |
| [NotificationResource.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5290/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.web.rest/NotificationResource.html) | 100% | ✅ |
| [NotificationSettingsResource.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5290/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.web.rest/NotificationSettingsResource.html) | 87% | ✅ |

### Screenshots
